app/uk/gov/hmrc/apiplatformevents/repository/ApplicationEventsRepository.scala

```diff
@@ -42,36 +40,25 @@ class ApplicationEventsRepository @Inject()(mongoComponent: ReactiveMongoCompone
       ReactiveMongoFormatters.formatApplicationEvent,
       ReactiveMongoFormats.objectIdFormats) {
 
+  override def indexes = Seq(
+    Index(
+      key = Seq("id" -> IndexType.Ascending),
+      name = Some("id_index"),
+      unique = true,
+      background = true
+    )
+  )
+
   def createEntity(event: ApplicationEvent): Future[Boolean] =
     insert(event).map(wr => wr.ok)
 
-  def populateEventIds(): Future[Unit] = {
-    val query : JsObject = Json.obj("id" -> Json.obj("$exists" -> false))
-
-    def updateApplicationEvent(event: JsObject): Unit = {
-      val mongoId = (event \ "_id").as[BSONObjectID]
-      val queryForFindById = Json.obj("_id" -> mongoId)
-      val setIdOnEvent = Json.obj("$set" -> Json.obj("id" -> UUID.randomUUID()))
-      val f : Future[_] = findAndUpdate(query = queryForFindById , setIdOnEvent)
-      f.onComplete {
-        case Failure(e) => Logger.error("Failed to update event" + e.getMessage)
-        case Success(_) => Logger.info(s"Success - Id added to event with _id: $mongoId")
-      }
-    }
-
-    processAll(query, updateApplicationEvent)
-  }
-
-  private def processAll(query : JsObject,  function: JsObject => Unit): Future[Unit] = {
-    import akka.stream.scaladsl.{Sink, Source}
-    import reactivemongo.akkastream.{State, cursorProducer}
-
-    val sourceOfEvents: Source[JsObject, Future[State]] =
-      collection
-        .find(query, Option.empty[JsObject])
-        .cursor[JsObject]()
-        .documentSource()
-
-    sourceOfEvents.runWith(Sink.foreach(function)).map(_ => ())
+  def fetchEventsToNotify[A <: ApplicationEvent](eventType: EventType)(implicit formatter: OFormat[A]): Source[A, Future[Any]] = {
+    val builder = collection.BatchCommands.AggregationFramework
+    val pipeline = List(
+      builder.Match(Json.obj("eventType" -> eventType)),
+      builder.Lookup(from = "notifications", localField = "id", foreignField = "eventId", as = "notifications"),
+      builder.Match(Json.obj("notifications" -> Json.obj("$size" -> 0)))
```

Comment:
The events to notify are the ones for which we don't have any notifications recorded yet (either SENT or FAILED)